File: src/backend/x86_64.c
#include "x86_64.h"

static const char *x86_64_register_names[] = {
    [R_EAX] = "eax",
};

static int32_t x86_64_emit_ret(FILE *out)
{
    fprintf(out, "\tret\n");
    return 0;
}

static int32_t x86_64_emit_mov(codegen_mov_instruction_t *mov, FILE *out)
{
    if (mov->source.tag == O_IMMEDIATE && mov->destination.tag == O_REGISTER) {
        fprintf(out, "\tmovl $%d, %%%s\n", mov->source.immediate, x86_64_register_names[mov->destination.register_id]);
    } else {
        fprintf(stderr, "[Error]: Unsupported mov instruction\n");
        return ERR_UNSUPPORTED_MOV_INSTRUCTION;
    }

    return 0;
}

static int32_t x86_64_emit_instruction_queue(codegen_instruction_queue_t queue, FILE *out)
{ 
    for (size_t i = 0; i < queue.count; i++) {
        codegen_instruction_t instruction = queue.instructions[i];
        int32_t err = 0;
        switch (instruction.tag) {
            case I_RET: {
                err = x86_64_emit_ret(out);
                break;
            };
            case I_MOV: {
                codegen_mov_instruction_t *mov = (codegen_mov_instruction_t *) instruction.data;
                err = x86_64_emit_mov(mov, out);
                break;
            };
        }

        if (err) {
            return err;
        }
    }

    return 0;
}

static int32_t x86_64_emit_function(codegen_function_definition_t function, FILE *out)
{
    // Emit label, in macOS, the label should start with an underscore
#ifdef __APPLE__
    fprintf(out, ".globl _%s\n", function.identifier);
    fprintf(out, "_%s:\n", function.identifier);
#else
    fprintf(out, ".globl %s\n", function.identifier);
    fprintf(out, "%s:\n", function.identifier);
#endif

    int32_t err = x86_64_emit_instruction_queue(function.instructions, out);
    if (err) {
        return err;
    }

// on linux we must add .section .note.GNU-stack,"",@progbits to avoid stack execution
#ifdef __linux__
    fprintf(out, ".section .note.GNU-stack,\"\",@progbits\n");
#endif

    return 0;
}

int32_t x86_64_codegen(arena_t *arena, char *buf, size_t size, const char *output_path)
{
    codegen_program_t program = {0};
    int32_t err = codegen_whole_program(arena, buf, size, &program);

    if (err) {
        return err;
    }

    FILE *out = fopen(output_path, "w");
    if (!out) {
        fprintf(stderr, "[Error]: Could not create output file\n");
        return ERR_COULD_NOT_CREATE_OUTPUT_FILE;
    }

    err = x86_64_emit_function(program.function, out);
    fclose(out);

    return 0;
}

File: src/backend/x86_64.zig
pub const std = @import("std");
const builtin = @import("builtin");
pub const codegen = @import("../codegen.zig");

pub const AssemblyGenerationError = error{
    InvalidRegister,
    FailedToCreateOutputFile,
    FaileToWriteOutputFile,
    InvalidMovAdressingMode,
};

pub fn getRegisterByName(reg: u64) AssemblyGenerationError![]const u8 {
    switch (reg) {
        0 => return "eax",
        else => return AssemblyGenerationError.InvalidRegister,
    }
}

pub const X86_64 = struct {
    const Self = @This();
    allocator: std.mem.Allocator,

    pub fn x86_64GenAssembly(self: *Self, outputPath: []const u8, program: codegen.Program) AssemblyGenerationError!void {
        // First, let's open the file
        const file = std.fs.cwd().createFile(outputPath, .{}) catch {
            return AssemblyGenerationError.FailedToCreateOutputFile;
        };
        defer file.close();

        try self.x86_64EmitFunction(file.writer(), program.functionDefinition);
    }

    fn x86_64EmitFunction(self: *Self, writer: std.fs.File.Writer, function: codegen.FunctionDefinition) AssemblyGenerationError!void {
        const symbolPrefix: []const u8 = if (comptime builtin.os.tag == .macos) "_" else "";

        // Emit Label
        writer.print(".globl {s}{s}\n", .{ symbolPrefix, function.identifier }) catch {
            return AssemblyGenerationError.FaileToWriteOutputFile;
        };

        writer.print("{s}{s}:\n", .{ symbolPrefix, function.identifier }) catch {
            return AssemblyGenerationError.FaileToWriteOutputFile;
        };

        try self.x86_64EmitInstructions(writer, function.instructions);
    }

    fn x86_64EmitInstructions(self: *Self, writer: std.fs.File.Writer, instructions: codegen.InstructionList) AssemblyGenerationError!void {
        defer instructions.deinit();

        for (instructions.items) |instruction| {
            switch (instruction.tag) {
                codegen.Instruction.Tag.ret => {
                    try x86_64EmitRetInstruction(writer);
                    self.allocator.destroy(instruction.value.retInstruction);
                },
                codegen.Instruction.Tag.mov => {
                    try x86_64EmitMovInstruction(writer, instruction.value.moveInstruction.*);
                    self.allocator.destroy(instruction.value.moveInstruction);
                },
            }
        }
    }

    fn x86_64EmitRetInstruction(writer: std.fs.File.Writer) AssemblyGenerationError!void {
        writer.print("\tret\n", .{}) catch {
            return AssemblyGenerationError.FaileToWriteOutputFile;
        };
    }

    fn x86_64EmitMovInstruction(writer: std.fs.File.Writer, instruction: codegen.MovInstruction) AssemblyGenerationError!void {
        if (instruction.src.tag == codegen.Operand.Tag.immediate and instruction.dst.tag == codegen.Operand.Tag.register) {
            const register = try getRegisterByName(instruction.dst.value.register);
            writer.print("\tmovl ${d}, %{s}\n", .{ instruction.src.value.immediate, register }) catch {
                return AssemblyGenerationError.FaileToWriteOutputFile;
            };
        } else {
            return AssemblyGenerationError.InvalidMovAdressingMode;
        }
    }
};
